src/sentry/replays/consumers/recording/process_recording.py

```diff
@@ -173,48 +174,65 @@ def submit(self, message: Message[KafkaPayload]) -> None:
                 "Failed to process replay recording message", extra={"offset": message.offset}
             )
 
-    def join(self, timeout: Optional[float] = None) -> None:
-        wait([f for _, f in self.__futures], timeout=timeout, return_when=ALL_COMPLETED)
-        self._commit_and_prune_futures()
-
     def close(self) -> None:
         self.__closed = True
 
-    def _commit_and_prune_futures(self) -> None:
-        """
-        Commit the latest offset of any completed message from the original
-        consumer.
-        """
+    def terminate(self) -> None:
+        self.close()
+        self.__threadpool.shutdown(wait=False)
 
-        committable: MutableMapping[Partition, Message[KafkaPayload]] = {}
+    def join(self, timeout: Optional[float] = None) -> None:
+        start = time.time()
+
+        # Immediately commit all the offsets we have popped from the queue.
+        self.__throttled_commit(force=True)
+
+        # Any remaining items in the queue are flushed until the process is terminated.
+        while self.__futures:
+            remaining = timeout - (time.time() - start) if timeout is not None else None
+            if remaining is not None and remaining <= 0:
+                logger.warning(f"Timed out with {len(self.__futures)} futures in queue")
+                break
+
+            # Pop the future from the queue.  If it succeeds great but if not it will be discarded
+            # on the next loop iteration without commit.  An error will be logged.
+            message, future = self.__futures.popleft()
+
+            try:
+                future.result(remaining)
+                self.__commit({message.partition: Position(message.offset, message.timestamp)})
```

Comment:
would we want to commit regardless of an error here? not sure if it matters too much.

Reply:
The way snuba handles it is to not commit and then die (leaving the queue full of items).  Not sure whats best to be honest.

Reply:
cc @lynnagara

Reply:
Talked with lyn.  This commit behavior is fine.  We don't want to commit failed messages.  Should be able to retry.

Reply:
sounds good. 🚢